lifecycle-hooks.md

```diff
@@ -0,0 +1,188 @@
+```
+
+> **Optional**  
+> If you are curious, take a look at `.status.handlers` of the `printer-extension` resource by doing `kubectl get extensionconfig printer-extension -o jsonpath='{.status.handlers}'`. This lists down all the lifecycle events the extension server would like to receive.
+
+## Extension in Action
+
+Now that we have the extension server running and registered with Cluster API let's see it in action. The extension server is configured to receive the following events (non-exhaustive list):
+
+- BeforeClusterCreate
+- AfterControlPlaneInitialized
+- BeforeClusterDelete
+
+### Create a new workload Cluster
+
+In this section we will create a new workload cluster and see that the extension server receives these events and logs them.  
+In a new terminal tab let's start watching the logs of the extension server:
+
+```bash
+kubectl logs -n test-extension-system deployments/test-extension-manager -f
+```
+
+Go back to the previous tab and create a new `docker-cluster-lifecycle-hooks` workload cluster.
+```bash
+kubectl apply -f yamls/clusters/docker-cluster-lifecycle-hooks.yaml
+```
+
+In the extension server you should see logs similar to:
+
+```bash
+...... other log lines
+I1014 02:50:50.860237       1 handlers.go:47] "BeforeClusterCreate is called"
+I1014 02:50:50.873831       1 handlers.go:165] "BeforeClusterCreate response is Success. retry: 0"
+I1014 02:51:38.782571       1 handlers.go:80] "AfterControlPlaneInitialized is called"
+```
+
+Note that the workload cluster could take a few minutes to fully come up. Each of the log lines above would take a few minutes in between to show up.
+
+### Delete the Cluster
+
+In this section we will delete the `docker-cluster-lifecycle-hooks` cluster and observe that the extension server is receiving and logging the event.
+
+> If you are feeling a adventurous try the [(Optional) Block Cluster deletion using Extension Server](#optional-block-cluster-deletion-using-extension-server) to block the deleting of a cluster using extension.
+
+To delete the cluster run:
+
+```bash
+kubectl delete cluster docker-cluster-lifecycle-hooks
+```
+
+In the extension server you should see logs similar to:
+
+```bash
+I1014 03:06:52.393668       1 handlers.go:131] "BeforeClusterDelete is called"
+I1014 03:06:52.408447       1 handlers.go:165] "BeforeClusterDelete response is Success. retry: 0"
+```
+
+### (Optional) Block Cluster deletion using Extension Server
+
+Till now we used a simple extension server that only logs the lifecycle events. What if we wanted to do more with the lifecycle hooks? In this section we will see how lifecycle hooks can be used to do more than just log events. 
+
+Let's change our extension server's behavior to send a "blocking response" when it receives a `BeforeClusterDelete` event.
+
+The extension server is pre-configured to always send a "allow response". Let's change this so that the extension server now send a "blocking response" to the `BeforeClusterDelete` event.
```

Comment:
```suggestion
The extension server is pre-configured to always send a "allow response". Let's change this so that the extension server now sends a "blocking response" to the `BeforeClusterDelete` event.
```